qa/tasks/rsync.py

```diff
@@ -0,0 +1,285 @@
+import logging
+import contextlib
+import time
+from teuthology import misc
+from teuthology.exceptions import CommandFailedError
+from gevent.greenlet import Greenlet
+from gevent.event import Event
+from tasks.cephfs.filesystem import Filesystem
+
+log = logging.getLogger(__name__)
+
+'''
+    parameters:
+
+    filesize: Optional Param. Unit in MB, default file size 2MB
+              Size of the data file before creating snapshot.
+              Not of use if workunit task running in parallel.
+
+    runtime: Required Param for using rsync task sequentially. unit in seconds. default 0 sec.
+             Don't use this param during parallel execution. if used then finally block will be delayed
+
+    data_dir: Don't use this param if not rsyncing data/snap from workunit data
+              Optional Param. Specify the client like client.0/client.1
+              make sure cephfs mount and workunit task using this client according to scenario.
+              Default source directory will be source/subdir for other scenarios.
+
+    snapenable: Optional Param. value in boolean (True or False), default False(disabled).
+                using this option will enable cephfs snapshot.
+
+    waittime: Optional Param. units in seconds. default 5 sec.
+              During rsync iteration each loop will wait for specified seconds.
+
+    mountpoint: Optional param. default it will be first ceph-fuse client id.
+            eg: if there are 3 cephfs clients then for client.0 use 0, client.1 use 1
+            Setting this option will run rsync on this particular client, make sure cephfs mount available.
+
+    Examples:
+    rsync task parallel with another task
+    tasks:
+      - ceph:
+      - ceph-fuse:
+      - rsync:
+            waittime: 10
+            snapenable: True
+            filesize: 8
+      - cephfs_test_runner:
+
+    rsync data from workunit directory in same client.
+    tasks:
+      - ceph:
+      - ceph-fuse: [client.2]
+      - rsync:
+            waittime: 10
+            mountpoint: 2
+            data_dir: client.2
+      - workunit:
+            clients:
+            client.2:
+                - suites/iozone.sh
+
+    rsync task parallel with workunit but on different client
+    tasks:
+      - ceph:
+      - ceph-fuse:
+      - rsync:
+            waittime: 10
+            mountpoint: 1
+      - workunit:
+            clients:
+            client.2:
+                - suites/iozone.sh
+
+    rsync task sequentially based on runtime. default runtime is 0
+    tasks:
+      - ceph:
+      - ceph-fuse:
+      - rsync:
+            runtime: 120
+            mountpoint: 2
+            snapenable: True
+
+'''
+
+class RSync(Greenlet):
```

Comment:
We could have inherit RSync from the Thrasher class, as it follows the same design as thrashers. But an `RSync` is not of a `Thrasher` type. So @batrick what do you suggest here?

Reply:
They are different. No reason to inherit.

However, the DaemonWatchdog should probably monitor it. Maybe you should generalize the thrasher monitoring in DaemonWatchdog.

Reply:
okay, I will do it as a separate patch. @sidharthanup Please add an exception variable for RSync then, so that DaemonWatchdog could check for exceptions.